### PR TITLE
PyObjectAlloc no longer uses specialization and freelist is removed

### PR DESCRIPTION
**Pull request by ijl**
The `impl<T> PyObjectAlloc` for PyObjectWithFreeList is an obstacle to
removing specialization. It is unused in all projects linked from
the README and within pyo3 itself, so let's just remove it.

### CODE REVIEW

**Comment by ijl** [comment]
cc #210

**Comment by konstin** [comment]
I have mixed feelings about this. It's true that it's an obstacle for compiling on stable and not widely used. OTOH it's a feature coming directly from the python c api and is relevant to performance. Would it be possible to feature gate with an off-by-default `freelist` feature?

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,10 +22,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
  * Slowly removing specialization uses
  * `PyString`, `PyUnicode`, and `PyBytes` no longer have a `data()` method
  (replaced by `as_bytes()`) and `PyStringData` has been removed.
+ * `pyo3::freelist` removed.
```
Comment (konstin) [inline CHANGELOG.md]: Now that 0.5.0 is released this should probably go in the unreleased block

**Comment by kngwyu** [comment]
I'm **against** this change.
To remove specialize, what we need to do is just to remove `impl<T:  PyObjectWithFreeList> PyObjectAlloc for T` and add `impl PyObjectAlloc for #cls` in proc_macro.
There's no reason to remove `freelist.rs` itself.

**Comment by kngwyu** [comment]
I'm against this change.
To remove default here means **we can't override alloc/dealloc** and I don't think it's a correct approach.
I think there can be some possible approaches here(e.g.  integrating PyObjectAlloc into PyTypeInfo / separating PyObjectWithFreeList from PyTypeInfo), and I don't think to remove freelist is the best one.

**Comment by konstin** [comment]
@kngwyu default doesn't work on stable, so we need to remove it someway. It's still possible to support overwrite alloc by using macro arguments.

@fafhrd91 you added the freelist; Do you know projects using that feature?

**Comment by kngwyu** [comment]
@konstin
But we have custom dealloc implementation here https://github.com/rust-numpy/rust-numpy/blob/master/src/slice_box.rs#L67
So how about moving PyObjectAlloc's methods into PyTypeInfo?
If so we can still add custom alloc/dealloc implementations for a type which implements PyTypeInfo.

**Comment by konstin** [comment]
What would you think about [this branch](https://github.com/PyO3/pyo3/compare/alloc_experiment)? Afaik that should work for all cases and we can even allow overriding with `#[pyclass]` later.

**Pull request closed without merging**